python/cudf_polars/cudf_polars/experimental/fusion.py

```diff
@@ -0,0 +1,187 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024-2025, NVIDIA CORPORATION & AFFILIATES.
+# SPDX-License-Identifier: Apache-2.0
+"""IR node fusion."""
+
+from __future__ import annotations
+
+import operator
+from collections import defaultdict
+from functools import reduce
+from typing import TYPE_CHECKING, Any
+
+from cudf_polars.dsl.ir import (
+    IR,
+    Cache,
+    Filter,
+    GroupBy,
+    HStack,
+    MapFunction,
+    Projection,
+    Select,
+    Union,
+)
+from cudf_polars.dsl.traversal import CachingVisitor, traversal
+from cudf_polars.experimental.dispatch import generate_ir_tasks
+from cudf_polars.experimental.io import Scan, SplitScan
+
+if TYPE_CHECKING:
+    from collections.abc import Callable, MutableMapping, Sequence
+    from typing import Any
+
+    from cudf_polars.containers import DataFrame
+    from cudf_polars.experimental.base import PartitionInfo
+    from cudf_polars.experimental.dispatch import LowerIRTransformer
+    from cudf_polars.typing import Schema
+
+
+class Fused(IR):
+    """Fused IR node."""
+
+    __slots__ = ("fused_io", "subnodes")
+    _non_child = ("schema", "subnodes", "fused_io")
+    subnodes: tuple[IR, ...]
+    """Fused sub-nodes."""
+    fused_io: Union | None
+    """Fused IO node."""
+
+    def __init__(
+        self,
+        schema: Schema,
+        subnodes: tuple[IR, ...],
+        fused_io: Union | None,
+        *children: IR,
+    ):
+        self.schema = schema
+        self.subnodes = subnodes
+        self.fused_io = fused_io
```

Comment:
*question*: It seems like `fused_io` is never used in the `do_evaluate`. How does this work? Ah, it's because the task generation does the right thing.

Does that suggest we should have separate `Fused` and `FusedIO` nodes? So we don't have this dichotomy?

Reply:
>Does that suggest we should have separate Fused and FusedIO nodes? So we don't have this dichotomy?

FYI: I refactored things with this goal in mind, and cleaned things up a bit in the process.